File: androidApp/src/main/java/dev/xinto/argos/ui/screen/course/page/syllabus/SyllabusPage.kt
package dev.xinto.argos.ui.screen.course.page.syllabus

import androidx.annotation.StringRes
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.unit.dp
import androidx.core.os.bundleOf
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import dev.xinto.argos.R
import dev.xinto.argos.ui.component.MaterialHtmlText2
import org.koin.androidx.compose.getStateViewModel

@Composable
fun SyllabusPage(
    courseId: String,
    modifier: Modifier = Modifier
) {
    val viewModel: SyllabusViewModel = getStateViewModel(state = {
        bundleOf(SyllabusViewModel.KEY_COURSE_ID to courseId)
    })
    val state by viewModel.state.collectAsStateWithLifecycle()
    SyllabusPage(
        state = state,
        modifier = modifier
    )
}

@Composable
fun SyllabusPage(
    state: SyllabusState,
    modifier: Modifier = Modifier
) {
    Box(
        modifier = modifier,
        contentAlignment = Alignment.Center
    ) {
        when (state) {
            is SyllabusState.Loading -> {
                CircularProgressIndicator()
            }

            is SyllabusState.Success -> {
                val context = LocalContext.current
                val text = remember(context) {
                    buildString {
                        fun append(@StringRes title: Int, body: String?) {
                            if (body != null) {
                                append("<div style=\"display:flex; flex-direction:column; gap:8px;\">")
                                append("<h3>")
                                append(context.getString(title))
                                append("</h3>")
//                                    appendLine()
                                append(body)
                                append("</div>")
                            }
                        }
                        append("<div style=\"display:flex; flex-direction:column; gap:12px;\">")
                        append(R.string.syllabus_section_duration, state.syllabus.duration)
                        append(R.string.syllabus_section_hours, state.syllabus.hours)
                        append(R.string.syllabus_section_lecturers, state.syllabus.lecturers)
                        append(
                            R.string.syllabus_section_prerequisites,
                            state.syllabus.prerequisites
                        )
                        append(R.string.syllabus_section_methods, state.syllabus.methods)
                        append(R.string.syllabus_section_mission, state.syllabus.mission)
                        append(R.string.syllabus_section_topics, state.syllabus.topics)
                        append(R.string.syllabus_section_outcomes, state.syllabus.outcomes)
                        append(R.string.syllabus_section_evaluation, state.syllabus.evaluation)
                        append(R.string.syllabus_section_resources, state.syllabus.resources)
                        append(
                            R.string.syllabus_section_resources_other,
                            state.syllabus.otherResources
                        )
                        append(R.string.syllabus_section_schedule, state.syllabus.schedule)
                        append("</div>")
                    }
                }
                MaterialHtmlText2(
                    text = text,
                    modifier = Modifier.fillMaxSize(),
                    contentPadding = PaddingValues(16.dp)
                )
            }

            is SyllabusState.Error -> {
                Text(text = "Error")
            }
        }
    }
}

File: iosApp/Argos/Course Screen/SyllabusPage.swift
//
//  SyllabusPage.swift
//  iosApp
//
//  Created by Tornike Khintibidze on 25.09.24.
//  Copyright © 2024 orgName. All rights reserved.
//

import SwiftUI
import ArgosCore
import Observation

@Observable
class SyllabusViewModel {
    
    @ObservationIgnored private var fetchTask: Task<Void, Error>?
    @ObservationIgnored private var courseRepository: CoursesRepository {
        DiProvider.shared.coursesRepository
    }
    
    var state: SyllabusState = .loading
    
    init(courseId: String) {
        fetchTask = Task {
            for try await syllabusData in courseRepository.getCourseSyllabus(courseId: courseId).flow {
                switch onEnum(of: syllabusData) {
                case .loading:
                    state = .loading
                case .success(let syllabus):
                    state = .success(syllabus.value!)
                case .error:
                    state = .error
                }
            }
        }
    }
    
    deinit {
        fetchTask?.cancel()
    }
}

struct SyllabusPage: View {
    
    private let viewModel: SyllabusViewModel
    
    init(courseId: String) {
        viewModel = SyllabusViewModel(courseId: courseId)
    }
    
    var body: some View {
        _SyllabusPage(state: viewModel.state)
    }
}

struct _SyllabusPage: View {
    
    let state: SyllabusState
    
    private var syllabus: String {
        guard case .success(let data) = state else {
            return ""
        }
        
        var syllabus = String()

        func append(section: String, body: String?) {
            guard let body else {
                return
            }
            
            syllabus.append("<div style=\"display:flex; flex-direction:column; gap:8px;\">")
            syllabus.append("<h3>")
            syllabus.append(section)
            syllabus.append("</h3>")
            syllabus.append(body)
            syllabus.append("</div>")
        }
        
        syllabus.append("<div style=\"display:flex; flex-direction:column; gap:12px;\">")
        append(section: "Duration", body: data.duration)
        append(section: "Hours", body: data.hours)
        append(section: "Lecturers", body: data.lecturers)
        append(section: "Prerequisites", body: data.prerequisites)
        append(section: "Methods", body: data.methods)
        append(section: "Mission", body: data.mission)
        append(section: "Topics", body: data.topics)
        append(section: "Outcomes", body: data.outcomes)
        append(section: "Evaluation", body: data.evaluation)
        append(section: "Resources", body: data.resources)
        append(section: "Other resources", body: data.otherResources)
        append(section: "Schedule", body: data.schedule)
        syllabus.append("</div>")
        
        return syllabus
    }
    
    var body: some View {
        switch state {
        case .loading:
            ProgressView()
        case .success:
            HtmlText2(syllabus)
                .htmlText2ContentPadding(12)
        case .error:
            Text("Error")
        }
    }
}

enum SyllabusState {
    case loading
    case success(DomainCourseSyllabus)
    case error
}
